scripts/ci/upload-r2-object.py

```
@@ -104,6 +105,44 @@ def _build_signed_request(
+    attempts = int(os.environ.get("CMUX_R2_UPLOAD_MAX_ATTEMPTS", "5"))
+    delay = float(os.environ.get("CMUX_R2_UPLOAD_RETRY_DELAY_SECONDS", "1"))
+    return max(1, attempts), max(0.0, delay)
+
+
+def _open_with_retries(request: urllib.request.Request, *, timeout: float, what: str):
+    """urlopen that retries R2's transient answers; the last failure propagates."""
+
+    attempts, delay = _retry_settings()
+    for attempt in range(1, attempts + 1):
+        try:
+            return urllib.request.urlopen(request, timeout=timeout)
```

Ruff (0.16.3)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)